server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -34,58 +44,253 @@
     this.authenticator = new MyAuthenticator(clientDAO, clientDAO);
+    this.clientAuthConfig = clientAuthConfig;
   }
 
   public Client provide(ContainerRequest request) {
-    Optional<Principal> principal = getPrincipal(request);
-    Optional<String> possibleClientName = getClientName(principal);
-    if (!principal.isPresent() || !possibleClientName.isPresent()) {
-      throw new NotAuthorizedException("ClientCert not authorized as a Client");
+    // Ports must either always send an x-forwarded-client-cert header, or
+    // never send this header. This also throws an error if a single port
+    // has multiple configurations.
+    int requestPort = request.getBaseUri().getPort();
+    Optional<XfccSourceConfig> possibleXfccConfig =
+        getXfccConfigForPort(requestPort);
+
+    List<String> xfccHeaderValues =
+        Optional.ofNullable(request.getRequestHeader(XFCC_HEADER_NAME)).orElse(List.of());
+
+    if (possibleXfccConfig.isEmpty() != xfccHeaderValues.isEmpty()) {
+      throw new NotAuthorizedException(format(
+          "Port %d is configured to %s receive traffic with the %s header set",
+          requestPort, possibleXfccConfig.isEmpty() ? "never" : "only", XFCC_HEADER_NAME));
     }
-    String clientName = possibleClientName.get();
 
-    return authenticator.authenticate(clientName, principal.orElse(null))
-        .orElseThrow(() -> new NotAuthorizedException(
-            format("ClientCert name %s not authorized as a Client", clientName)));
+    // Extract information about the requester. This may be a Keywhiz client, or it may be a proxy
+    // forwarding the real Keywhiz client information in the x-forwarded-client-certs header
+    Principal requestPrincipal = getPrincipal(request).orElseThrow(
+        () -> new NotAuthorizedException("Not authorized as Keywhiz client"));
+
+    Optional<String> requestName = getClientName(requestPrincipal);
+    Optional<String> requestSpiffeId = Optional.empty();
+
+    // Extract client information based on the x-forwarded-client-cert header or
+    // on the security context of this request
+    if (possibleXfccConfig.isEmpty()) {
+      // The XFCC header is not used; use the security context of this request to identify the client
+      return authorizeClient(requestPrincipal, requestName, requestSpiffeId);
+    } else {
+      return authorizeClientFromXfccHeader(possibleXfccConfig.get(), xfccHeaderValues, requestName,
+          requestSpiffeId);
+    }
   }
 
   static Optional<Principal> getPrincipal(ContainerRequest request) {
     return Optional.ofNullable(request.getSecurityContext().getUserPrincipal());
   }
 
-  static Optional<String> getClientName(Optional<Principal> principal) {
-    if (!principal.isPresent()) {
-      return Optional.empty();
-    }
-
-    X500Name name = new X500Name(principal.get().getName());
+  static Optional<String> getClientName(Principal principal) {
```

Comment:
I suspect it's because `request.getSecurityContext().getUserPrincipal()` for a `ContainerRequest` returns only a `Principal`, so specifying an `X500Principal` here would require casting. 

(I haven't scoped how much work would be involved in rewriting `AuthResolver` to provide something other than a `ContainerRequest` to this factory.)